### PR TITLE
Rest proxy enhanced error responses

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes issue #3302 

Allows repeating UnexpectedResponseExceptionType annotations on methods.

### CODE REVIEW

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionTypes.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * have different exceptions to be thrown or returned based on the response status codes returned from a REST API.
+ */
+@Retention(RUNTIME)
+@Target(METHOD)
+public @interface UnexpectedResponseExceptionTypes {
+    /**
+     * @return array of {@link UnexpectedResponseExceptionType} that annotate a method.
+     */
+    UnexpectedResponseExceptionType[] value();
+}
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionTypes.java]: I don't believe this annotation is required.

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java [hunk 2]
```diff
@@ -1,37 +1,51 @@
...
+
+ * {@literal @}UnexpectedResponseExceptionType(code = {404}, value = NotFoundException.class)
+ * {@literal @}UnexpectedResponseExceptionType(DefaultException.class)
+ * {@literal @}POST("subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/images/getEntityTypeImageUploadUrl")
+ *  void getUploadUrlForEntityType(@Path("resourceGroupName") String resourceGroupName, @Path("hubName") String hubName, @Path("subscriptionId") String subscriptionId, @Body GetImageUploadUrlInputInner parameters);
  * </pre>
  */
 @Retention(RUNTIME)
 @Target(METHOD)
+@Repeatable(UnexpectedResponseExceptionTypes.class)
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java]: This could probably just be the singular type `UnexpectedResponseExceptionType`

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java [hunk 3]
```diff
@@ -1,37 +1,51 @@
...
 
 import static java.lang.annotation.ElementType.METHOD;
 import static java.lang.annotation.RetentionPolicy.RUNTIME;
 
 /**
- * The error type that will be thrown or returned when an unexpected status code is returned from an REST API.
+ * The exception type that is thrown or returned when one of the status codes is returned from a REST API. Multiple
+ * annotations can be used, they are processed in a top-down order. When no codes are listed that exception is always
+ * thrown or returned if it is reached during evaluation, this should be treated as a default case. If no default case
+ * is annotated the fall through exception is ServiceRequestException.
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java]: Don't fully agree on this approach. I would look up specifically for a stated status code. If that is not found, then there should only be one annotation specified without a code, so you use that (if there are more than one without a code specified, you should throw an exception)

File: common/azure-common/src/main/java/com/azure/common/implementation/RestProxy.java [hunk 4]
```diff
@@ -252,22 +247,18 @@ private HttpRequest configRequest(HttpRequest request, SwaggerMethodParser metho
...
-            }
-            else if (bodyContentObject instanceof String) {
+            } else if (bodyContentObject instanceof String) {
                 final String bodyContentString = (String) bodyContentObject;
                 if (!bodyContentString.isEmpty()) {
                     request.withBody(bodyContentString);
                 }
-            }
-            else {
+            } else {
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/RestProxy.java]: You should avoid the desire to make unrelated changes when submitting a PR. These kind of formatting changes should ideally not be included in a PR for a new feature.

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java [hunk 5]
```diff
@@ -1,14 +1,8 @@
-/**
- * Copyright (c) Microsoft Corporation. All rights reserved.
- * Licensed under the MIT License. See License.txt in the project root for
- * license information.
- */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: You are probably going to have merge errors due to changes made by Sima here - another reason to only touch what is relevant.

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java [hunk 6]
```diff
@@ -149,21 +140,18 @@ else if (TypeUtil.isTypeOrSubTypeOf(returnValueWireType, List.class)) {
...
-            exceptionType = unexpectedResponseExceptionType.value();
+        for (UnexpectedResponseExceptionType exceptionAnnotation : swaggerMethod.getAnnotationsByType(UnexpectedResponseExceptionType.class)) {
+            UnexpectedException exception = new UnexpectedException(exceptionAnnotation.value());
+            if (exceptionAnnotation.code().length == 0) {
+                exceptionMapping.putIfAbsent(FALLTHROUGH_EXCEPTION_CODE, exception);
+            } else {
+                for (int statusCode : exceptionAnnotation.code()) {
+                    exceptionMapping.putIfAbsent(statusCode, exception);
+                }
+            }
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Is it necessary to do this work? Could it be delayed until it is needed (i.e when an exception is needed to be thrown)?

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java [hunk 7]
```diff
@@ -48,6 +46,7 @@
  * method.
  */
 public class SwaggerMethodParser implements HttpResponseDecodeData {
+    private static final int FALLTHROUGH_EXCEPTION_CODE = -1;
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: This isn't a good approach. If a match isn't found in a map, just fall through - don't put something arbitrary into a map just to then look it up in a situation you know is exceptional.

File: common/azure-common/src/main/java/com/azure/common/implementation/RestProxy.java (hunk 4 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/RestProxy.java]: Reverted these types of changes out, I left the import sorting and removing changes though

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java (hunk 5 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Reverted

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java [hunk 8]
```diff
@@ -526,4 +519,31 @@ private String applySubstitutions(String originalValue, Iterable<Substitution> s
 
         return result;
     }
+
+    private HashMap<Integer, UnexpectedException> processUnexpectedResponseExceptionTypes() {
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Return the `Map` interface, not a concrete implementation

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java (hunk 7 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Added a defaultException member variable that will be used in place of this.

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java (hunk 6 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Construction of the mapping is now done when the first exception response is seen.

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java [hunk 9]
```diff
@@ -526,4 +519,31 @@ private String applySubstitutions(String originalValue, Iterable<Substitution> s
...
+                    throw new InvalidUnexpectedResponseAnnotationsException(fullyQualifiedMethodName);
+                }
+            } else {
+                for (int statusCode : exceptionAnnotation.code()) {
+                    if (exceptionHashMap.putIfAbsent(statusCode, exception) != null) {
+                        throw new InvalidUnexpectedResponseAnnotationsException(fullyQualifiedMethodName, statusCode);
+                    }
+                }
+            }
+        }
```
Comment (JonathanGiles) [inline common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: I think we should change this. Don't throw exceptions in either case - just take the most recent value as the definitive one and replace the previous one.

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java (hunk 3 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java]: Updated how the implementation works to throw an exception when multiple default exceptions are seen or when a HTTP status code is shared across annotations, update the JavaDoc to reflect that.

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java (hunk 2 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionType.java]: This is required as the Repeatable annotation needs a class that has a value method that returns an array of the class its annotating, in this case `UnexpectedResponseExceptionType[] value()`.

File: common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionTypes.java (hunk 1 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/annotations/UnexpectedResponseExceptionTypes.java]: This is required as the Repeatable annotation needs a class that has a value method that returns an array of the class its annotating, in this case `UnexpectedResponseExceptionType[] value()`.

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java (hunk 8 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Changed to map

File: common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java (hunk 9 above)
Comment (alzimmermsft) [reply common/azure-common/src/main/java/com/azure/common/implementation/SwaggerMethodParser.java]: Removed exception throws
